README.md
[ ![Become a Patreon](http://larsnorbergofficial.com/img/social-media-buttons-patreon-small.jpg) ](https://www.patreon.com/goldpawsstuff) 

This is a custom user interface suite for World of Warcraft, containing multiple full interfaces in a single addon. Coded and designed by Lars Norberg and Daniel Troconis. It works for both WoW Retail and WoW Classic, just choose the right version to download!  

To switch between interface theme, use the commands `/go azerite` and `/go legacy`. This is a per character setting.  

## **Motivate me!**  
If you like this addon or others I have made, and wish to see more, consider supporting my work by sending a donation with PayPal or pledging monthly on Patreon. Your support is what motivates me to share my addons with the world, develop them further and keep them updated! 

* Patreon: [www.patreon.com/goldpawsstuff](https://www.patreon.com/goldpawsstuff)  
* PayPal: [www.paypal.me/goldpawsstuff](https://www.paypal.me/goldpawsstuff)  
* Liberapay: [liberapay.com/goldpawsstuff/donate](https://liberapay.com/goldpawsstuff/donate)

### **Join me on social media!**  
* Discord: [discord.gg/MUSfWXd](https://discord.gg/MUSfWXd)  
* Twitter: [@GoldpawsStuff](https://twitter.com/goldpawsstuff)  

### **Report a bug, suggest a feature:**
All feature requests and bug reports for all game versions are done on the main project's github page. Please use _only_ our report system and the supplied template when reporting bugs.  
* [github.com/GoldpawsStuff/AzeriteUI/issues](https://github.com/GoldpawsStuff/AzeriteUI/issues)

## **Recommended addons:**  
These are some addons I feel go very well together with our user interface. They are not required, but I recommend them as they further enhance the look and functionality in a non-intrusive manner.  

### **Retail:**
* Big Input Box: [www.curseforge.com/wow/addons/big-input-box](https://www.curseforge.com/wow/addons/big-input-box)
* Immersion: [www.curseforge.com/wow/addons/immersion](https://www.curseforge.com/wow/addons/immersion)  
* Narcissus: [www.curseforge.com/wow/addons/narcissus](https://www.curseforge.com/wow/addons/narcissus)  
* MaxDps: [www.curseforge.com/wow/addons/maxdps-rotation-helper](https://www.curseforge.com/wow/addons/maxdps-rotation-helper)
* Minimap Button Bag: [www.curseforge.com/wow/addons/mbb](https://www.curseforge.com/wow/addons/mbb)  
* World Map Shrinker: [www.curseforge.com/wow/addons/world-map-shrinker](https://www.curseforge.com/wow/addons/world-map-shrinker)

### **Classic Era:**
* Big Input Box: [www.curseforge.com/wow/addons/big-input-box](https://www.curseforge.com/wow/addons/big-input-box)
* Enhanced World Map for WoW Classic: [www.curseforge.com/wow/addons/enhanced-world-map-for-wow-classic](https://www.curseforge.com/wow/addons/enhanced-world-map-for-wow-classic)  
* Immersion for WoW Classic: [www.curseforge.com/wow/addons/immersion](https://www.curseforge.com/wow/addons/immersion/files/all?filter-game-version=1738749986%3a67408)  

### **Burning Crusade Classic:**
* Big Input Box: [www.curseforge.com/wow/addons/big-input-box](https://www.curseforge.com/wow/addons/big-input-box)
* Enhanced World Map for WoW Classic: [www.curseforge.com/wow/addons/enhanced-world-map-for-wow-classic](https://www.curseforge.com/wow/addons/enhanced-world-map-for-wow-classic)  
* Immersion for WoW Classic: [www.curseforge.com/wow/addons/immersion](https://www.curseforge.com/wow/addons/immersion/files/all?filter-game-version=1738749986%3a67408)  
* TotemTimers Classic: [www.curseforge.com/wow/addons/totemtimers-classic](https://www.curseforge.com/wow/addons/totemtimers-classic)  

## **Frequently Asked Questions:**  
### **It's so big, how can I scale it down?**  
* You can't. It's meant to be this big! We're a couple of lazy, laid back gamers, and prefer gaming in a fair distance from our screens, keeping things relaxed and simplified. Big health bars, easy to see big buttons, nothing there we don't absolutely need. We treat this as a console game, and this UI is built around that philosophy. The sizes should thus be relatively the same on all screens, relative to the height of the screen. 

### **How can I move things?**  
* You can't. This is as much UX as UI, meaning the interface is designed around a concept, this isn't meant to be a pretty skin over your current interface, it's instead a full user experience system where how you interact with the game is affected. 

### **How can I get more action buttons, and where is the stance bar?**  
* For more buttons, right-click the cogwheel icon in the bottom right corner of the screen. You can have a maximum of 2 full bars in the azerite theme, and a maximum of 4 full bars in the legacy theme.
* The stance bar will be introduced as an optional feature in an upcoming update. If you need it right away, you can always use Bartender for it along with our [our Masque skin](https://www.curseforge.com/wow/addons/masque-azerite). 

### **How can I disable the actionbars?**
* This is not a feature I intend to implement. I do realize that some of you would like this feature for it to look better with ConsolePort, or just completely replace our bars with Bartender, but dismantling our user interface to make it fit a completely different project slightly better just isn't something we're going to do. 


CHANGELOG.md
# AzeriteUI Change Log
All notable changes to this project will be documented in this file. Be aware that the [Unreleased] features are not yet available in the official tagged builds.

The format is based on [Keep a Changelog](http://keepachangelog.com/) 
and this project adheres to [Semantic Versioning](http://semver.org/).

## [3.2.497-RC] 2021-06-10
### Fixed
- Fixed the weird and sudden issue with the QuestTimerFrame in BC and Classic. 

## [3.2.496-RC] 2021-06-09
### Added
- There should be a focus frame in BC clients now. 

## [3.2.495-RC] 2021-05-31
- New tag to re-trigger GitHub actions and get builds uploaded to the now functioning CurseForge. 

### Added
- Added the minimap tracking button to BC clients, and also the minimap right-click functionality as an additional way to bring up the BC client tracking menu.
- Added the nicely styled bottom button to the chat frame in BC clients.
- Attempted to add the pet happiness element to BC clients. Untested, as I have no hunter with pet there yet.

### Changed
- Moved the quest timer frame used in classic quests such as Iverron's Antidote to the same annoying place in BC clients as in classic.

## [3.2.493-RC] 2021-05-28
### Added
- Added the option to completely disable player- and target auras when using the Legacy theme. 

## [3.2.492-RC] 2021-05-26
### Added
- Added the option to completely disable player- and target auras when using the standard Azerite theme. 

## [3.2.491-RC] 2021-05-24
### Changed
- Increased max nameplate distance in BC to 30 yards outdoors, and 41 yards in instances. 

## Fixed
- Hovering over an NPC with no questie global options yet saved should no longer cause a bug in bc or classic.

## [3.2.490-RC] 2021-05-23
### Fixed
- Various annoying micro menu alerts should now longer cause randomly hovering alert windows.

## [3.2.489-RC] 2021-05-23
### Fixed
- The collections journal should no longer cause an addon action blocked when opened during combat.

## [3.2.488-RC] 2021-05-23
### Added
- Attempted to add the same battleground finder eye to the BC minimap as we used in Classic. Untested.
- Added Questie information to BC tooltips and nameplates.

## [3.2.487-RC] 2021-05-22
### Fixed
- Added the same workaround for the tainted blizzard battleground popups in Burning Crusade as we used in Classic.

## [3.2.486-RC] 2021-05-20
### Fixed
- The experience bar should now be visible on Burning Crusade realms.

## [3.2.485-RC] 2021-05-17
- Extra push needed because the bigwigs packager changed its API from using "bc" to calling it "bcc". 

## [3.2.484-RC] 2021-05-17
### Fixed
- You should no longer get two identical target frames in legacy mode.

## [3.2.483-RC] 2021-05-13
### Changed
- WorldFrame unit tooltips no longer get their content modified by this addon if totalRP3 currently is enabled. 

## [3.2.482-RC] 2021-05-08
- First batch of fixes for BC compatibility.

## [3.2.480-Release] 2021-05-01
### Fixed
- Fixed a typo in the legacy theme unitframe module that would cause party frames to be hidden on startup and reloads until manually toggled through the menu. 

## [3.2.479-Release] 2021-04-29
### Added
- Added blip icons for WoW client patch 9.0.5.

### Fixed
- Certain toybox items with a sellprice even though they're not actually sellable will no longer cause a bug if you hover the mouse cursor above them on the action bar while holding down the shift key. 

## [3.2.478-Release] 2021-04-27
### Changed
- The temporary weapon enchant buttons (shaman weapon buffs, classic fishing lures, etc) will now follow the opacity of the player unit frame. It also has a new place to live up at the minimap in the legacy theme.
- Unit tooltips where the creature's family or type can't be determined will now mimick the default tooltip behavior and display no info about this, instead of the rather meaningless text "Not specified".

## [3.2.477-Release] 2021-04-26
### Fixed
- The focus frame which got lost a few updates back has found its way back home.

## [3.2.476-Release] 2021-04-22
### Fixed
- Fixed an issue that would break the UI for multiple classes when a tutorial was shown during the Exile's Reach intro campaign.

## [3.2.475-Release] 2021-04-20
- Minor bug fixes and workarounds.

## [3.2.474-Release] 2021-04-19
### Added
- Added menu options to toggle the Legacy theme's party- and raid frames! Now you can use Legacy with your favorite group frame addon!

### Changed
- Gave the Blizzard popups a slight facelift. They are now darker and more transparent, and just allround more beautiful. Like me. 
- There should now be a short delay from entering the Minimap's number badge until the ring frames showing the full XP/Rep/Azerite bars appear. This should help avoid that annoying situation where you accidentally touch the badge on your way to the minimap only to get the map covered by bars while you were trying to hover above an object or player on the map.   

## [3.2.473-RC] 2021-04-15
- Crazy changes.

## [3.2.470-RC] 2021-04-14
### Changed
- Updated the MaxDps integration, hopefully to the better. Keeping an eye on this.

### Fixed
- Fixed an issue in Classic that would leave an empty border around the class power element on some classes that didn't have a class power element. 

## [3.2.469-RC] 2021-04-11
### Added
- Added support for WorldMapShrinker!

### Changed
- Moving anything chat filter related out of the back-end. This is part of an ongoing chat filter overhaul. 

AzeriteUI/back-end/plugins/unit_targethighlight.lua
local LibClientBuild = Wheel("LibClientBuild")
assert(LibClientBuild, "UnitThreat requires LibClientBuild to be loaded.")

-- Lua API
local string_match = string.match

-- WoW API
local UnitIsUnit = UnitIsUnit

-- Constants for client version
local IsRetail = LibClientBuild:IsRetail()
local IsTBC = LibClientBuild:IsTBC()

local Update = function(self, event, unit, ...)
	if (not unit) or (unit ~= self.unit) then 
		return 
	end 
	local element = self.TargetHighlight
	if (element.PreUpdate) then
		element:PreUpdate(unit)
	end

	-- Hide if the owner is too transparent
	if (not element.ignoreAlpha) and (self:GetAlpha() < .1) then 
		element:Hide()

	-- Don't highlight the focus frame as the current focus
	elseif (element.showFocus) and (unit ~= "focus") and (UnitIsUnit("focus", unit)) then 
		if (element.colorFocus) then 
			element:SetVertexColor(element.colorFocus[1], element.colorFocus[2], element.colorFocus[3], element.colorFocus[4])
		end
		element:Show()

	-- Don't highlight the target frame nor the tot frame as the current target
	elseif (element.showTarget) and (unit ~= "target") and (unit ~= "targettarget") and (UnitIsUnit("target", unit)) then 
		if (element.colorTarget) then 
			element:SetVertexColor(element.colorTarget[1], element.colorTarget[2], element.colorTarget[3], element.colorTarget[4])
		end
		element:Show()
	else
		element:Hide()
	end

	if (element.PostUpdate) then 
		return element:PostUpdate(unit)
	end
end 

local Proxy = function(self, ...)
	return (self.TargetHighlight.Override or Update)(self, ...)
end 

local ForceUpdate = function(element)
	return Proxy(element._owner, "Forced", element._owner.unit)
end

local Enable = function(self)
	local element = self.TargetHighlight
	if (element) then
		element._owner = self
		element.ForceUpdate = ForceUpdate

		self:RegisterEvent("PLAYER_TARGET_CHANGED", Proxy, true)

		if (string_match(self.unit, "^party(%d+)")) or (string_match(self.unit, "^raid(%d+)")) then
			self:RegisterEvent("GROUP_ROSTER_UPDATE", Proxy, true)
		end

		if (IsRetail or IsTBC) then
			self:RegisterEvent("PLAYER_FOCUS_CHANGED", Proxy, true)
		end

		return true 
	end
end 

local Disable = function(self)
	local element = self.TargetHighlight
	if (element) then

		self:UnregisterEvent("GROUP_ROSTER_UPDATE")
		self:UnregisterEvent("PLAYER_TARGET_CHANGED")
		self:UnregisterEvent("PLAYER_FOCUS_CHANGED")

		element:Hide()
	end
end 

-- Register it with compatible libraries
for _,Lib in ipairs({ (Wheel("LibUnitFrame", true)), (Wheel("LibNamePlate", true)) }) do 
	Lib:RegisterElement("TargetHighlight", Enable, Disable, Proxy, 14)
end 


AzeriteUI/back-end/libraries/unitframe.lua
local LibUnitFrame = Wheel:Set("LibUnitFrame", 89)
if (not LibUnitFrame) then	
	return
end

local LibEvent = Wheel("LibEvent")
assert(LibEvent, "LibUnitFrame requires LibEvent to be loaded.")

local LibFrame = Wheel("LibFrame")
assert(LibFrame, "LibUnitFrame requires LibFrame to be loaded.")

local LibClientBuild = Wheel("LibClientBuild")
assert(LibClientBuild, "LibSecureButton requires LibClientBuild to be loaded.")

local LibWidgetContainer = Wheel("LibWidgetContainer")
assert(LibWidgetContainer, "LibUnitFrame requires LibWidgetContainer to be loaded.")

local LibTooltip = Wheel("LibTooltip")
assert(LibTooltip, "LibUnitFrame requires LibTooltip to be loaded.")

local LibSound = Wheel("LibSound")
assert(LibSound, "LibUnitFrame requires LibSound to be loaded.")

LibEvent:Embed(LibUnitFrame)
LibFrame:Embed(LibUnitFrame)
LibTooltip:Embed(LibUnitFrame)
LibWidgetContainer:Embed(LibUnitFrame)

-- Lua API
local _G = _G
local assert = assert
local debugstack = debugstack
local error = error
local math_floor = math.floor
local pairs = pairs
local select = select
local setmetatable = setmetatable
local string_format = string.format
local string_gsub = string.gsub
local string_join = string.join
local string_match = string.match
local table_insert = table.insert
local table_remove = table.remove
local tonumber = tonumber
local type = type
local unpack = unpack

-- Blizzard API
local CreateFrame = CreateFrame
local FriendsDropDown = FriendsDropDown
local SecureCmdOptionParse = SecureCmdOptionParse
local ShowBossFrameWhenUninteractable = ShowBossFrameWhenUninteractable
local ToggleDropDownMenu = ToggleDropDownMenu
local UnitExists = UnitExists
local UnitGUID = UnitGUID
local UnitHasVehicleUI = UnitHasVehicleUI
local UnitIsEnemy = UnitIsEnemy
local UnitIsFriend = UnitIsFriend

-- Constants for client version
local IsClassic = LibClientBuild:IsClassic()
local IsTBC = LibClientBuild:IsTBC()
local IsRetail = LibClientBuild:IsRetail()

-- Library Registries
LibUnitFrame.embeds = LibUnitFrame.embeds or {} -- who embeds this?
LibUnitFrame.frames = LibUnitFrame.frames or  {} -- global unitframe registry
LibUnitFrame.scriptHandlers = LibUnitFrame.scriptHandlers or {} -- tracked library script handlers
LibUnitFrame.scriptFrame = LibUnitFrame.scriptFrame -- library script frame, will be created on demand later on

-- Speed shortcuts
local frames = LibUnitFrame.frames
local elements = LibUnitFrame.elements
local callbacks = LibUnitFrame.callbacks
local unitEvents = LibUnitFrame.unitEvents
local frequentUpdates = LibUnitFrame.frequentUpdates
local frequentUpdateFrames = LibUnitFrame.frequentUpdateFrames
local frameElements = LibUnitFrame.frameElements
local frameElementsEnabled = LibUnitFrame.frameElementsEnabled
local scriptHandlers = LibUnitFrame.scriptHandlers
local scriptFrame = LibUnitFrame.scriptFrame

-- Color Table
--------------------------------------------------------------------------
-- RGB to Hex Color Code
local hex = function(r, g, b)
	return ("|cff%02x%02x%02x"):format(math_floor(r*255), math_floor(g*255), math_floor(b*255))
end

-- Convert a Blizzard Color or RGB value set 
-- into our own custom color table format. 
local prepare = function(...)
	local tbl
	if (select("#", ...) == 1) then
		local old = ...
		if (old.r) then 
			tbl = {}
			tbl[1] = old.r or 1
			tbl[2] = old.g or 1
			tbl[3] = old.b or 1
		else
			tbl = { unpack(old) }
		end
	else
		tbl = { ... }
	end
	if (#tbl == 3) then
		tbl.colorCode = hex(unpack(tbl))
	end
	return tbl
end

-- Convert a whole Blizzard color table
local prepareGroup = function(group)
	local tbl = {}
	for i,v in pairs(group) do 
		tbl[i] = prepare(v)
	end 
	return tbl
end 

-- Default Color Table
local Colors = {
	artifact = prepare( 229/255, 204/255, 127/255 ),
	class = prepareGroup(RAID_CLASS_COLORS),
	dead = prepare( 153/255, 153/255, 153/255 ),
	debuff = prepareGroup(DebuffTypeColor),
	disconnected = prepare( 153/255, 153/255, 153/255 ),
	feedback = {
		DAMAGE 			= prepare( 176/255,  79/255,  79/255 ),
		CRUSHING 		= prepare( 176/255,  79/255,  79/255 ),
		CRITICAL 		= prepare( 176/255,  79/255,  79/255 ),
		GLANCING 		= prepare( 176/255,  79/255,  79/255 ),
		STANDARD 		= prepare( 214/255, 191/255, 165/255 ),
		IMMUNE 			= prepare( 214/255, 191/255, 165/255 ),
		ABSORB 			= prepare( 214/255, 191/255, 165/255 ),
		BLOCK 			= prepare( 214/255, 191/255, 165/255 ),
		RESIST 			= prepare( 214/255, 191/255, 165/255 ),
		MISS 			= prepare( 214/255, 191/255, 165/255 ),
		HEAL 			= prepare(  84/255, 150/255,  84/255 ),
		CRITHEAL 		= prepare(  84/255, 150/255,  84/255 ),
		ENERGIZE 		= prepare(  79/255, 114/255, 160/255 ),
		CRITENERGIZE 	= prepare(  79/255, 114/255, 160/255 )
	},
	health = prepare( 25/255, 178/255, 25/255 ),
	power = { ALTERNATE = prepare(70/255, 255/255, 131/255) },
	quest = {
		red = prepare( 204/255, 25/255, 25/255 ),
		orange = prepare( 255/255, 128/255, 25/255 ),
		yellow = prepare( 255/255, 204/255, 25/255 ),
		green = prepare( 25/255, 178/255, 25/255 ),
		gray = prepare( 153/255, 153/255, 153/255 )
	},
	reaction = prepareGroup(FACTION_BAR_COLORS),
	rested = prepare( 23/255, 93/255, 180/255 ),
	restedbonus = prepare( 192/255, 111/255, 255/255 ),
	tapped = prepare( 153/255, 153/255, 153/255 ),
	xp = prepare( 18/255, 179/255, 21/255 )
}

-- Adding this for semantic reasons, 
-- so that plugins can use it for friendly players
-- and the modules will have the choice of overriding it.
Colors.reaction.civilian = Colors.reaction[5]

-- Power bar colors need special handling, 
-- as some of them contain sub tables.
for powerType, powerColor in pairs(PowerBarColor) do 
	if (type(powerType) == "string") then 
		if (powerColor.r) then 
			Colors.power[powerType] = prepare(powerColor)
		else 
			if powerColor[1] and (type(powerColor[1]) == "table") then 
				Colors.power[powerType] = prepareGroup(powerColor)
			end 
		end  
	end 
end 

-- Secure Snippets
--------------------------------------------------------------------------
local secureSnippets = {

}

-- Utility Functions
--------------------------------------------------------------------------
-- Syntax check 
local check = function(value, num, ...)
	assert(type(num) == "number", ("Bad argument #%.0f to '%s': %s expected, got %s"):format(2, "Check", "number", type(num)))
	for i = 1,select("#", ...) do
		if type(value) == select(i, ...) then 
			return 
		end
	end
	local types = string_join(", ", ...)
	local name = string_match(debugstack(2, 2, 0), ": in function [`<](.-)['>]")
	error(string_format("Bad argument #%.0f to '%s': %s expected, got %s", num, name, types, type(value)), 3)
end

-- Library Updates
--------------------------------------------------------------------------
-- global update limit, no elements can go above this
local THROTTLE = 1/30 
local OnUpdate = function(self, elapsed)

	-- Throttle the updates, to increase the performance. 
	self.elapsed = (self.elapsed or 0) + elapsed
	if (self.elapsed < THROTTLE) then
		return
	end
	local elapsed = self.elapsed

	for frame, frequentElements in pairs(frequentUpdates) do
		for element, frequency in pairs(frequentElements) do
			if frequency.hz then
				frequency.elapsed = frequency.elapsed + elapsed
				if (frequency.elapsed >= frequency.hz) then
					elements[element].Update(frame, "FrequentUpdate", frame.unit, elapsed) 
					frequency.elapsed = 0
				end
			else
				elements[element].Update(frame, "FrequentUpdate", frame.unit)
			end
		end
	end

	self.elapsed = 0
end

-- Unitframe Template
--------------------------------------------------------------------------
local UnitFrame = {} 
local UnitFrame_MT = { __index = UnitFrame }

-- Return or create the library default tooltip
-- This is shared by all unitframes, unless these methods 
-- are specifically overwritten by the modules.
UnitFrame.GetTooltip = function(self)
	return LibUnitFrame:GetUnitFrameTooltip()
end 

UnitFrame.OnEnter = function(self)
	if ((not self.unit) or (not UnitExists(self.unit))) then 
		return 
	end 

	self.isMouseOver = true

	local tooltip = self:GetTooltip()
	tooltip:Hide()
	tooltip:SetDefaultAnchor(self)
	tooltip:SetMinimumWidth(160)
	tooltip:SetUnit(self.unit)

	if self.PostEnter then 
		self:PostEnter()
	end 
end

UnitFrame.OnLeave = function(self)
	self.isMouseOver = nil

	local tooltip = self:GetTooltip()
	tooltip:Hide()

	if self.PostLeave then 
		self:PostLeave()
	end 
end

UnitFrame.OnHide = function(self)
	self.unitGUID = nil
end

UnitFrame.OverrideAllElements = function(self, event, ...)
	local unit = self.unit
	if (not unit) or (not UnitExists(unit) and not ShowBossFrameWhenUninteractable(unit)) then 
		return 
	end
	if (self.isMouseOver) then
		local OnEnter = self:GetScript("OnEnter")
		if (OnEnter) then
			OnEnter(self)
		end
	end
	return self:UpdateAllElements(event, ...)
end

-- Special method that only updates the elements if the GUID has changed. 
-- Intention is to avoid performance drops from people coming and going in PuG raids. 
UnitFrame.OverrideAllElementsOnChangedGUID = function(self, event, ...)
	local unit = self.unit
	if (not unit) or (not UnitExists(unit) and not ShowBossFrameWhenUninteractable(unit)) then 
		return 
	end
	local currentGUID = UnitGUID(unit)
	if currentGUID and (self.unitGUID ~= currentGUID) then 
		self.unitGUID = currentGUID
		if (self.isMouseOver) then
			local OnEnter = self:GetScript("OnEnter")
			if (OnEnter) then
				OnEnter(self)
			end
		end
		if (unit == "target") and (not self.noTargetChangeSoundFX) then
			if (UnitExists("target")) then
				-- Play a fitting sound depending on what kind of target we gained
				if (UnitIsEnemy("target", "player")) then
					LibSound:PlaySoundKitID(SOUNDKIT.IG_CREATURE_AGGRO_SELECT, "SFX")
				elseif (UnitIsFriend("player", "target")) then
					LibSound:PlaySoundKitID(SOUNDKIT.IG_CHARACTER_NPC_SELECT, "SFX")
				else
					LibSound:PlaySoundKitID(SOUNDKIT.IG_CREATURE_NEUTRAL_SELECT, "SFX")
				end
			else
				-- Play a sound indicating we lost our target
				LibSound:PlaySoundKitID(SOUNDKIT.INTERFACE_SOUND_LOST_TARGET_UNIT, "SFX")
			end
		end
		return self:UpdateAllElements(event, ...)
	end
end

local UpdatePet = function(self, event, unit)
	local petUnit
	if (unit == "target") then
		return
	elseif (unit == "player") then
		petUnit = "pet"
	else
		petUnit = unit.."pet"
	end
	if (not self:OnAttributeChanged("unit", UnitHasVehicleUI(unit) and petUnit or unit)) then
		return self:UpdateAllElements(event, "Forced", self.unit)
	end
end

-- Library API
--------------------------------------------------------------------------
-- Return or create the library default tooltip
LibUnitFrame.GetUnitFrameTooltip = function(self)
	return LibUnitFrame:GetTooltip("GP_UnitFrameTooltip") or LibUnitFrame:CreateTooltip("GP_UnitFrameTooltip")
end

LibUnitFrame.SetScript = function(self, scriptHandler, script)
	scriptHandlers[scriptHandler] = script
	if (scriptHandler == "OnUpdate") then
		if (not scriptFrame) then
			scriptFrame = CreateFrame("Frame", nil, LibFrame:GetFrame())
		end
		if script then 
			scriptFrame:SetScript("OnUpdate", function(self, ...) 
				script(LibUnitFrame, ...) 
			end)
		else
			scriptFrame:SetScript("OnUpdate", nil)
		end
	end
end

LibUnitFrame.GetScript = function(self, scriptHandler)
	return scriptHandlers[scriptHandler]
end

LibUnitFrame.GetUnitFrameVisibilityDriver = function(self, unit, hideInVehicles)
	local visDriver
	if (unit == "player") then
		if (IsClassic or IsTBC) then
			visDriver = "[@player,exists][mounted]show;hide"
		elseif (IsRetail) then
			-- Might seem stupid, but I want the player frame to disappear along with the actionbars 
			-- when we've blown the flight master's whistle and are getting picked up.
			if (hideInVehicles) then
				visDriver = "[vehicleui]hide;[@player,exists][possessbar][overridebar][mounted]show;hide"
			else
				visDriver = "[@player,exists][vehicleui][possessbar][overridebar][mounted]show;hide"
			end
		end
	elseif (unit == "pet") then
		if (IsRetail) then
			-- Adding this to avoid situations where the player frame is hidden, 
			-- yet a pet frame hovers in mid-air.
			-- This only happens in short periods, like when a quest requires you to 
			-- take a flight on a dragon to fly to or return to someplace.
			-- Still, looks silly. We want it fixed.
			local prefix = "[@player,noexists]hide;"
			if (hideInVehicles) then
				visDriver = prefix .. "[vehicleui]hide;[@pet,exists]show;hide"
			else
				visDriver = prefix .. "[@pet,exists][nooverridebar,vehicleui]show;hide"
			end
		end
	else
		local partyID = string_match(unit, "^party(%d+)")
		if (partyID) then
			if (hideInVehicles) then
				visDriver = string_format("[vehicleui]hide;[nogroup:raid,@%s,exists]show;hide", unit)
			else
				visDriver = string_format("[nogroup:raid,@%s,exists]show;hide", unit)
			end
		end
	end
	if (visDriver) then
		return visDriver
	else
		if (hideInVehicles) then
			return string_format("[vehicleui]hide;[@%s,exists]show;hide", unit)
		else
			return string_format("[@%s,exists]show;hide", unit)
		end
	end
end 

LibUnitFrame.GetUnitFrameUnitDriver = function(self, unit)
	local unitDriver
	if (IsRetail) then
		if (unit == "player") then 
			-- Should work in all cases where the unitframe is replaced. It should always be the "pet" unit.
			--unitDriver = "[vehicleui]pet;player"
			unitDriver = "[nooverridebar,vehicleui]pet;[overridebar,@vehicle,exists]vehicle;player"
		elseif (unit == "pet") then 
			unitDriver = "[nooverridebar,vehicleui]player;pet"
		elseif (string_match(unit, "^party(%d+)")) then 
			unitDriver = string_format("[unithasvehicleui,@%s]%s;%s", unit, unit.."pet", unit)
		elseif (string_match(unit, "^raid(%d+)")) then 
			unitDriver = string_format("[unithasvehicleui,@%s]%s;%s", unit, unit.."pet", unit)
		end
	end
	return unitDriver
end 

-- spawn and style a new unitframe
LibUnitFrame.SpawnUnitFrame = function(self, unit, parent, styleFunc, ...)
	check(unit, 1, "string")
	check(parent, 2, "table", "string", "nil")
	check(styleFunc, 3, "function", "string", "nil")

	-- Alllow modules to use methods as styling functions. 
	-- We don't want to allow this in the widgetcontainer back-end,  
	-- so we need a bit of trickery here to make it happen. 
	if (type(styleFunc) == "string") then 
		local func = self[styleFunc]
		if func then 
			local module, method = self, styleFunc
			styleFunc = function(...) 
				-- Always call the method by name, 
				-- don't assume the function is the same each time. 
				-- Even though it is. So this is weird. 
				return module[method](self, ...)
			end 
		end 
	end 
	
	local frame = LibUnitFrame:CreateWidgetContainer("Button", parent, "SecureUnitButtonTemplate", unit, styleFunc, ...)
	for method,func in pairs(UnitFrame) do 
		frame[method] = func
	end 

	frame.requireUnit = true
	frame.colors = frame.colors or Colors
	frame.ignoredEvents = {}

	if (frame.ignoreMouseOver) then 
		frame:EnableMouse(false)
		frame:RegisterForClicks("")
	else 
		frame:SetAttribute("*type1", "target")
		frame:SetAttribute("*type2", "togglemenu")
		frame:SetScript("OnEnter", UnitFrame.OnEnter)
		frame:SetScript("OnLeave", UnitFrame.OnLeave)
		frame:RegisterForClicks("AnyUp")
	end 

	frame:SetScript("OnHide", UnitFrame.OnHide)

	local OverrideAllElements = UnitFrame.OverrideAllElementsOnChangedGUID -- UnitFrame.OverrideAllElements

	if (unit == "target") then
		frame:RegisterEvent("PLAYER_TARGET_CHANGED", OverrideAllElements, true)
		frame.ignoredEvents["PLAYER_TARGET_CHANGED"] = true

	elseif (unit == "focus") then
		if (IsRetail or IsTBC) then
			frame:RegisterEvent("PLAYER_FOCUS_CHANGED", OverrideAllElements, true)
			frame.ignoredEvents["PLAYER_FOCUS_CHANGED"] = true
		end

	elseif (unit == "mouseover") then
		frame:RegisterEvent("UPDATE_MOUSEOVER_UNIT", OverrideAllElements, true)
		frame.ignoredEvents["UPDATE_MOUSEOVER_UNIT"] = true

	elseif (unit:match("^arena(%d+)")) then
		frame.unitGroup = "arena"
		frame:SetFrameStrata("MEDIUM")
		frame:SetFrameLevel(1000)
		frame:RegisterEvent("ARENA_OPPONENT_UPDATE", OverrideAllElements, true)
		frame.ignoredEvents["ARENA_OPPONENT_UPDATE"] = true

	elseif (string_match(unit, "^boss(%d+)")) then
		frame.unitGroup = "boss"
		frame:SetFrameStrata("MEDIUM")
		frame:SetFrameLevel(1000)

		frame:RegisterEvent("INSTANCE_ENCOUNTER_ENGAGE_UNIT", OverrideAllElements, true)
		frame.ignoredEvents["INSTANCE_ENCOUNTER_ENGAGE_UNIT"] = true

		frame:RegisterEvent("UNIT_TARGETABLE_CHANGED", OverrideAllElements, true)
		frame.ignoredEvents["UNIT_TARGETABLE_CHANGED"] = true

	elseif (string_match(unit, "^party(%d+)")) then 
		frame.unitGroup = "party"
		frame:RegisterEvent("GROUP_ROSTER_UPDATE", OverrideAllElements, true)
		frame.ignoredEvents["GROUP_ROSTER_UPDATE"] = true

		if (IsRetail) then
			frame:RegisterEvent("UNIT_PET", UpdatePet)
		end

	elseif (string_match(unit, "^raid(%d+)")) then 
		frame.unitGroup = "raid"
		frame:RegisterEvent("GROUP_ROSTER_UPDATE", UnitFrame.OverrideAllElementsOnChangedGUID, true)
		frame.ignoredEvents["GROUP_ROSTER_UPDATE"] = true

		if (IsRetail) then
			frame:RegisterEvent("UNIT_PET", UpdatePet)
		end

	elseif (unit == "targettarget") then
		-- Need an extra override event here so the ToT frame won't appear to lag behind on target changes.
		frame:RegisterEvent("PLAYER_TARGET_CHANGED", OverrideAllElements, true)
		frame.ignoredEvents["PLAYER_TARGET_CHANGED"] = true
		frame:EnableFrameFrequent(.5, "unit")

	elseif (string_match(unit, "%w+target")) then
		frame:EnableFrameFrequent(.5, "unit")
	end

	frame:SetAttribute("unit", unit)

	local unitDriver = frame.unitOverrideDriver or LibUnitFrame:GetUnitFrameUnitDriver(unit)
	if (unitDriver) and (not frame.hideInVehicles) then 
		local unitSwitcher = CreateFrame("Frame", nil, nil, "SecureHandlerAttributeTemplate")
		unitSwitcher:SetFrameRef("UnitFrame", frame)
		unitSwitcher:SetAttribute("unit", unit)
		unitSwitcher:SetAttribute("_onattributechanged", [=[
			local frame = self:GetFrameRef("UnitFrame"); 
			frame:SetAttribute("unit", value); 
		]=])
		frame.realUnit = unit
		frame:SetAttribute("unit", SecureCmdOptionParse(unitDriver))
		RegisterAttributeDriver(unitSwitcher, "state-vehicleswitch", unitDriver)
	else
		frame:SetAttribute("unit", unit)
	end 

	local visDriver = LibUnitFrame:GetUnitFrameVisibilityDriver(unit, frame.hideInVehicles)
	if (frame.visibilityOverrideDriver) then 
		visDriver = frame.visibilityOverrideDriver
	elseif (frame.visibilityPreDriver) then
		visDriver = frame.visibilityPreDriver .. visDriver
	end

	frame:SetAttribute("visibilityDriver", visDriver)
	RegisterAttributeDriver(frame, "state-visibility", visDriver)

	-- This and a global name is pretty much  
	-- the shortest route to Clique compatibility.
	_G.ClickCastFrames = ClickCastFrames or {}
	ClickCastFrames[frame] = true

	frames[frame] = true

	if (frame.PostCreate) then
		frame:PostCreate()
	end 
	
	return frame
end

-- Make this a proxy for development purposes
LibUnitFrame.RegisterElement = function(self, ...)
	LibWidgetContainer:RegisterElement(...)
end 

-- Module embedding
local embedMethods = {
	SpawnUnitFrame = true,
	GetUnitFrameVisibilityDriver = true, 
	GetUnitFrameTooltip = true
}

LibUnitFrame.Embed = function(self, target)
	for method in pairs(embedMethods) do
		target[method] = self[method]
	end
	self.embeds[target] = true
	return target
end

-- Upgrade existing embeds, if any
for target in pairs(LibUnitFrame.embeds) do
	LibUnitFrame:Embed(target)
end
